include/aspect/particle/property/interface.h

```diff
@@ -0,0 +1,355 @@
+
+          /**
+           * Update function. This function is called every timestep for
+           * every particle to update it's properties. It is obvious that
+           * this function is called a lot, so its code should be efficient.
+           */
+          virtual
+          void
+          update_particle (unsigned int &data_position,
+                           std::vector<double> &particle_properties,
+                           const Point<dim> &position,
+                           const Vector<double> &solution,
+                           const std::vector<Tensor<1,dim> > &gradients);
+
+          /**
+           * Returns a bool, which is false in the default implementation,
+           * telling the property manager that no update is needed. Every
+           * plugin that implements this function should return true. This
+           * saves considerable computation time in cases, when no plugin needs
+           * to update tracer properties over time.
+           */
+          virtual
+          UpdateTimeFlags
+          need_update ();
+
+          virtual
+          void
+          data_length(std::vector<unsigned int> &length) const = 0;
+
+          /**
+           * Set up the name information for the particle property
+           *
+           * @param [in,out] names Vector that contains the property name
+           */
+          virtual
+          void
+          data_names(std::vector<std::string> &names) const = 0;
+
+
+          /**
+           * Declare the parameters this class takes through input files.
+           * Derived classes should overload this function if they actually do
+           * take parameters; this class declares a fall-back function that
+           * does nothing, so that property classes that do not take any
+           * parameters do not have to do anything at all.
+           *
+           * This function is static (and needs to be static in derived
+           * classes) so that it can be called without creating actual objects
+           * (because declaring parameters happens before we read the input
+           * file and thus at a time when we don't even know yet which
+           * property objects we need).
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           * The default implementation in this class does nothing, so that
+           * derived classes that do not need any parameters do not need to
+           * implement it.
+           */
+          virtual
+          void
+          parse_parameters (ParameterHandler &prm);
+      };
+
+
+
+      /**
+       * Manager class of properties - This class sets the data of the particles
+       * and updates it over time if requested by the user selected properties
+       */
+      template <int dim>
+      class Manager : public SimulatorAccess<dim>
```

Comment:
The documentation of the class (and in particular of some of the member functions) seems to suggest that the class manages the properties of one particle (or of individual particles), but it really only manages the properties of the entire collection of particles. Can you go through the various member functions and see what needs to be adjusted?